compiler/rustc_next_trait_solver/src/canonicalizer.rs

```diff
@@ -79,7 +79,11 @@ impl<'a, D: SolverDelegate<Interner = I>, I: Interner> Canonicalizer<'a, D, I> {
             cache: Default::default(),
         };
 
-        let value = value.fold_with(&mut canonicalizer);
+        let value = if value.has_type_flags(TypeFlags::NEEDS_CANONICAL_NEXT_SOLVER) {
```

Comment:
make this an associated constant of `Canonicalizer` instead. Or just a free one in this file